docs/internal/git_best_practice.rst

```diff
@@ -0,0 +1,211 @@
+
+- `<https://www.atlassian.com/git>`_ good introduction, but dives deep too
+- `<https://help.github.com>`_ covers a lot of detail and github specifics
+
+
+Main Principles
+------------------
+
+Below are the main guiding principles, we will expand on that through out the document
+
+- Treat commit history as a form of documentation
+- Group related changes together
+- Avoid combining unrelated changes into one commit
+- Communicate intent and context of the change not just what changed
+- Maintain mostly linear history
+
+In the ideal world code is written once and is perfect from the start: there are
+no mistakes and all future uses are anticipated and catered for at the time of
+inception. In this perfect world every new commit is new code which adds new
+features. In the real world people make mistakes in the code, third party APIs
+change without any warning, code formatting is sometimes off, all these things
+stand in the way, all need to be fixed in the process of developing the new
+feature. While all these changes are necessary, they are still incidental to the
+core task of new feature development. We should try to put each incidental
+change into a separate commit. Separating "preparatory work" from "actual new
+feature" really helps with code review.
+
+Commit messages are an important communication channel, use them to communicate
+information that can not be easily extracted from the code: why was the change
+necessary, what that means for code that might depend on the code that got
+changed. Try to provide context for the change without being overly verbose.
+
+Git is a very flexible tool and allows for a broad range of workflows, changes
+can be happening in a number of parallel branches, those branches can then be
+*merged* in any order creating a rather non-linear history which can be hard to
+follow. We adopt a more constrained approach, commonly referred as "Feature
+Branch Workflow". In short this means that there is one main branch ``master``
+or ``develop``, no major work is done in the main branch, instead new features
+are developed in separate feature branches, those then get merged into the main
+branch when ready.
+
+
+Commit Message Format
+-----------------------------
+
+Good commit messages make collaboration easier. Commit message should consist of
+short summary line, 50 characters of less, followed by a blank line, followed by
+expanded description. It is fine to omit expanded description and the blank line
+before it if the change is small. Start your summary line with a capital letter,
+don't put full stop [.] at the end of the summary line.
+
+Many tools, GitHub including, expect this format and sticking to it makes
+inspecting history more accessible.
+
+.. note ::
+
+    Could be helpful to prescribe how summary line is to be phrased. Which do we
+    prefer?
+
+    - Fixed spelling mistake in the help message
+    - Fix spelling mistake in the help message
```

Comment:
The second one is the typical standard (imperative mood). Git itself uses it ("Merge branch..." rather than "Merged branch...") and all major projects that I know of.

Reply:
Agreed, will change to recommend imperative mood